MiniMap/Lock on Horde

lyonelz96 opened the pull request:
## What issue is this PR solving?
Closes: #38 
Closes: #32 

## Description of change

### Stuff Added
  - Added **MiniMap folder** in Assets containing all the scripts/textures related to the MiniMap
  - Added **MiniMap folder** in Prefabs containing all prefabs related to the MiniMap
  - Added **Shader folder** containing an **Outline shader**, to highlight the selected zombies when locking on a horde
  - Added the **Outline material** (Uses the outline shader) to the zombie Prefab as a second material on its Mesh Renderer
  - Added **unity-ui-extensions package**. I needed this to draw lines on the Unity Canvas (Can you believe there is no easy way of drawing lines on the canvas)
  - Added **ZombieHordeBlipCreator script** to **FlockCenter prefab**

### Features
  - Minimap should work with any nxn map (As long as its x,z coordinates are (0,0)
  - You can click on the minimap and the camera will move there
  - You have to double click a horde (Represented by a red dot) to focus on it

### Things that could use some polish (Need your feedback on this)
  - The quality of the minimap (Better lighting, different details)
    - Layers need to be set up properly so that the MiniMap camera only renders/takes pic of the "environment" at the beginning. If not, things that we don't want on the minimap will appear there
  - When the camera goes out of bounds (Either moves too far away or zooms out to far away) the viewport get's messed up on the minimap
     - This can be fixed by setting some kind of limit to where the camera can go/zoom out based on the size of the map
  - Position of the minimap
  - Exit state from the locked horde (Like pressing the ESC key or something)
     - You can leave the locked state by clicking anywhere else on the minimap
  - That's all I could think of for now, let me know for any other stuff you find
  

## How to reproduce the changes
  - There are 3 scenes I worked/tested things on:
     - MiniMap
     - Sandbox
     - Game

  - Essentially the process would be the same for every scene.

### Process
  0. Do nothing because it should be already setup for these scenes, but if new scene follow the next steps
  1. Drag MiniMapManager prefab to scene
  2. Give the MiniMapManager the reference of the Floor object of the world (Drag and drop in the inspector)

  - That's pretty much it
  - You can enable the CameraHandlerBlip (Dot on minimap representing the center of where the camera is looking at by going to MiniMapManager prefab, then MiniMapCanvas/MiniMap/CamerHandlerBlip and enable the Image component

  - Camera Movement stays the same
    - WASD to move
    - Hold Right Click on mouse and drag to route
    - Hold Drag Wheel down on mouse to pull and drag (Kinda like grabbing the world and pulling it, but really you're the one moving)
    - Drag Wheel to zoom in/out


## Screenshots or GIFs

### Floor Scale = 100
![Floor Scale 100](https://user-images.githubusercontent.com/43832900/78819611-80292900-79a4-11ea-9f91-16d9360976c3.gif)

### Floor Scale = 50
![Floor Scale 50](https://user-images.githubusercontent.com/43832900/78819626-84eddd00-79a4-11ea-8174-cc46b59b26ef.gif)


### Flock Blip Movement
![Flock Blip Movement](https://user-images.githubusercontent.com/43832900/78819877-eb72fb00-79a4-11ea-8746-4f3137f951f0.gif)

### Camera going where you click
![Click and Go](https://user-images.githubusercontent.com/43832900/78820061-31c85a00-79a5-11ea-865c-4a378c86f03b.gif)

### Locking in to Zombie Horde through clicking
![Click and Follow Horde](https://user-images.githubusercontent.com/43832900/78820485-e5314e80-79a5-11ea-93d2-f73b5a7a4da3.gif)

### Splitting Horde/Following multiple Blips
 ![Horde Splitting](https://user-images.githubusercontent.com/43832900/78820700-40634100-79a6-11ea-97e6-393c893d69dd.gif)

IrregularMeter committed:
Add MiniMap Folder and MapTexture Script
  - MapTexture Script takes a screenshot of the Environment

IrregularMeter committed:
Set map texture to Canvas

IrregularMeter committed:
Add miniMap Icon to follow GameObject

IrregularMeter committed:
Add Unity UI Extensions Package

IrregularMeter committed:
Add ViewPortBox in MiniMap

IrregularMeter committed:
Save scene

IrregularMeter committed:
Add CameraClickAndGo Script / Needs Refactor

IrregularMeter committed:
Add MiniMapManager and Helper / Refactor

IrregularMeter committed:
Refactor

IrregularMeter committed:
Add MouseButton Check for Camera Movement in Minimap

IrregularMeter committed:
Add MiniMapManager Prefab
  - Add MiniMap in SandBox Scene

IrregularMeter committed:
Adjust map pic size

IrregularMeter committed:
Add the following changes:
  - CalculateCenterOfMass script on the CenterOfMass object (Child of FlockCenter)
  - Add Scripts for Click and Focus on Horde
  - Add ZombieHordeBlipManager (Takes care of Deletion and Position Update)
  - Add ZombieHordeBlipCreator script

IrregularMeter committed:
Refactor

IrregularMeter committed:
Refactor

IrregularMeter committed:
Add Select and Deselect of Horde on Minimap

IrregularMeter committed:
Add Camera lock to Horde

IrregularMeter committed:
Add Outline Shader to Zombies

gnoriega27 committed:
rebase with master

gnoriega27 committed:
Duplicate

gnoriega27 committed:
fixed merge

IrregularMeter committed:
Add ZombieHordeBlipCreator script to FlockCenter prefab

IrregularMeter committed:
Add MiniMap to Game Scene

gnoriega27 reviewed (commented):
Everything works fine. My only comment would be that the "double click" on a horde shouldn't lock you into that horde, rather, it should simply move the camera there. In other words: "No camera lock". It's open to discussion.

gnoriega27 reviewed (changes requested):
Remove the "Examples" directory

ghost commented:
As a reminder to myself:

As discussed in the meeting I have to look at 2 things:
  1. Exit condition for the Horde
  2. Fix bug where you are locked to one horde and want to go to another one

IrregularMeter committed:
Delete Examples Folder from Unity UI Package

IrregularMeter committed:
Fix MiniMap Bug where trying to lock on another horde while being locked onto one initially didn't work

IrregularMeter committed:
Add free movement option when locked to horde
  - Add feature so that you only DeSelect the horde when you left click on the minimap

IrregularMeter committed:
Save scene

IrregularMeter committed:
Trying to fix merging of hordes

taramic42 committed:
Futsed around with horde splitting to make sure it highlights zombies properly

IrregularMeter committed:
Add Singleton For ZombieHordeBlips

IrregularMeter committed:
Apply MiniMapManager prefab change
